integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/DeviceState.java

```diff
@@ -0,0 +1,105 @@
+import org.opensmartgridplatform.oslp.Oslp;
+import org.opensmartgridplatform.shared.infra.jms.MessageType;
+
+import lombok.Getter;
+import lombok.Setter;
+
+public class DeviceState {
+
+    @Getter
+    private final String deviceUID;
+    @Getter @Setter
+    private Integer sequenceNumber = 0;
+
+    private static final Integer SEQUENCE_NUMBER_MAXIMUM = 65535;
+
+    private final ConcurrentMap<MessageType, ConcurrentLinkedQueue<Oslp.Message>> mockedResponsesMap = new ConcurrentHashMap<>();
+    private final ConcurrentMap<MessageType, ConcurrentLinkedQueue<Oslp.Message>> receivedRequestsMap = new ConcurrentHashMap<>();
+
+    public DeviceState(final String deviceUID) {
+        this.deviceUID = deviceUID;
+    }
+
+    public Oslp.Message getMockedResponse(final MessageType messageType) throws DeviceSimulatorException {
+        return this.getFromMap(this.mockedResponsesMap, messageType);
+    }
+
+    public Oslp.Message getReceivedRequest(final MessageType messageType) throws DeviceSimulatorException {
+        return this.getFromMap(this.receivedRequestsMap, messageType);
+    }
+
+    public void addMockedResponse(final MessageType messageType, final Oslp.Message message) {
+        this.getQueue(this.mockedResponsesMap, messageType).add(message);
+    }
+
+    public void addReceivedRequest(final MessageType messageType, final Oslp.Message message) {
+        this.getQueue(this.receivedRequestsMap, messageType).add(message);
+    }
+
+    public boolean hasMockedResponses(final MessageType messageType) {
+        return !this.getQueue(this.mockedResponsesMap, messageType).isEmpty();
+    }
+
+    public boolean hasReceivedRequests(final MessageType messageType) {
+        return !this.getQueue(this.receivedRequestsMap, messageType).isEmpty();
+    }
+
+    public void incrementSequenceNumber() {
+        int numberToAddToSequenceNumberValue = 1;
+
+        if (ScenarioContext.current().get(PlatformKeys.NUMBER_TO_ADD_TO_SEQUENCE_NUMBER) != null) {
+            final String numberToAddAsNextSequenceNumber = ScenarioContext.current()
+                    .get(PlatformKeys.NUMBER_TO_ADD_TO_SEQUENCE_NUMBER)
+                    .toString();
+            if (!numberToAddAsNextSequenceNumber.isEmpty()) {
+                numberToAddToSequenceNumberValue = Integer.parseInt(numberToAddAsNextSequenceNumber);
+            }
+        }
+        int next = this.sequenceNumber + numberToAddToSequenceNumberValue;
+        if (next > SEQUENCE_NUMBER_MAXIMUM) {
+            final int sequenceNumberMaximumCross = next - SEQUENCE_NUMBER_MAXIMUM;
+            if (sequenceNumberMaximumCross >= 1) {
+                next = sequenceNumberMaximumCross - 1;
+            }
+        } else if (next < 0) {
+            final int sequenceNumberMaximumCross = next * -1;
+            if (sequenceNumberMaximumCross >= 1) {
+                next = SEQUENCE_NUMBER_MAXIMUM - sequenceNumberMaximumCross + 1;
+            }
+        }
```

Comment:
In case that is the intended behaviour :-) After some simple refactoring I came to this if/else simplifcation:
        int next = this.sequenceNumber + increment;
        if (next > SEQUENCE_NUMBER_MAXIMUM) {
            next -= SEQUENCE_NUMBER_MAXIMUM - 1;
        } else if (next < 0) {
            next = SEQUENCE_NUMBER_MAXIMUM - next * -1 + 1;
        }
which is kinda weird... why an else? how could next become < 0? I guess only if you set a negative increment. But why would you do that? Is that possible?

Reply:
ProtocolSequenceNumber.feature covers tests where a negative increment is indeed possible.... The test does not cover edge cases. But for this reason I added the else again.